syncthing/XGSyncthing.m

```diff
@@ -37,25 +42,43 @@ - (void) stopExecutable
 
 - (id)sendRequestToEndpoint:(NSString *)endpoint method:(NSString *)method parameters:(NSDictionary *)parameters
 {
-    NSData *serverData = nil;
-    NSError *myError = nil;
-    NSURLResponse *serverResponse = nil;
-    NSMutableURLRequest *theRequest=[NSMutableURLRequest
-                                     requestWithURL:[NSURL URLWithString:[NSString stringWithFormat:@"%@%@", _URI, endpoint]]
-                                     cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval:10.0];
+    if (self.session == nil) {
+        // Lazy create a reasonable URLSession
+        NSURLSessionConfiguration *cfg = [NSURLSessionConfiguration defaultSessionConfiguration];
+        self.session = [NSURLSession sessionWithConfiguration:cfg delegate:[[LocalhostTLSDelegate alloc] init] delegateQueue:nil];
+    }
+
+    // Create the URL from base, endpoint and params.
+    NSURLComponents *comps = [NSURLComponents componentsWithString:[NSString stringWithFormat:@"%@%@", _URI, endpoint]];
+    NSMutableArray *queryItems = [NSMutableArray array];
+    for (NSString *key in parameters) {
+        [queryItems addObject:[NSURLQueryItem queryItemWithName:key value:[NSString stringWithFormat:@"%@", parameters[key]]]];
+    }
+    comps.queryItems = queryItems;
+
+    NSMutableURLRequest *theRequest = [NSMutableURLRequest
+                                       requestWithURL:[comps URL]
+                                       cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval:10.0];
     
     [theRequest setHTTPMethod:method];
     [theRequest setValue:_apiKey forHTTPHeaderField:@"X-API-Key"];
+    [theRequest setTimeoutInterval:EVENT_TIMEOUT + 5.0]; // slightly more than the Syncthing event timeout
     
-    serverData = [NSURLConnection sendSynchronousRequest:theRequest
-                                       returningResponse:&serverResponse error:&myError];
-    
-    if (myError)
-        return nil;
+    __block id result = nil;
+    dispatch_semaphore_t done = dispatch_semaphore_create(0);
```

Comment:
This stuff isn't exactly elegant. If there's a nicer way to un-async this I'm all ears. It's been years since I objective-c:d (objective-saw?) last.